rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,878 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 136 - Modern Signature Algorithms
+
+## Required Approvers
+
+* Engineering: (@jakule || @espadolini)
+* Security: (@reedloden || @jentfoo) && Doyensec
+* Product: @klizhentas
+
+## What
+
+This RFD proposes the adoption of a new set of asymmetric key types and
+signature algorithms across all Teleport protocols that surpass the 2048-bit RSA
+keys currently used in terms of security and performance.
+
+## Why
+
+Currently Teleport uses 2048-bit RSA keypairs for all internal Certificate
+Authorities and all user and host keys.
+
+More modern algorithms based on Elliptic Curve Cryptography offer better
+security properties with smaller keys that offer better performance for keypair
+generation and signing operations, with comparable performance for signature
+verification.  Some of the more restrictive security policies are starting to
+reject RSA2048 (e.g.  [RHEL 8's FUTURE policy](https://access.redhat.com/articles/3642912)).
+
+## Details
+
+### Summary
+
+We will introduce the concept of signature algorithm "suites" that will
+configure the key types and signature algorithms used across Teleport.
+This will be a cluster-wide configuration available in both the `teleport.yaml`
+config of Teleport Auth servers (so that initial CA keys generated at first
+startup can be configured) and in the `cluster_auth_preference` (so that Cloud
+users can optionally change the suite used in their cluster).
+
+The initial set of suites we will introduce is:
+
+* `legacy`
+* `balanced-v1`
+* `fips-v1`
+* `hsm-v1`
+
+The `legacy` suite exists for compatibility and it will use the exact same
+key types and algorithms Teleport uses today (2048-bit RSA everywhere).
+
+The `balanced-v1` suite will use a modern set of algorithms selected to
+balance security, compatibility, and performance.
+The proposed selection for this suite is to use Ed25519 as the default choice,
+and ECDSA with the NIST P-256 curve in cases where Teleport is likely to
+interact with third-party software that does not support Ed25519.
+The exception is the database CA, where it looks like it will be necessary to
+continue using RSA in order to continue supporting our current Snowflake
+integration.
```

Comment:
Will Snowflake be using RSA 2048 or a bigger key size?
From their documentation _"This authentication method requires, **as a minimum**, a 2048-bit RSA key pair. "_
https://docs.snowflake.com/en/user-guide/key-pair-auth

I suppose it is possible to use 3072-bit RSA keys to provide 128-bit security everywhere.

Reply:
Thanks for the review. 3072-bit RSA would probably be possible, but for now I think we are still okay with 2048-bit and would prefer to avoid the performance hit of switching to 3072, I added a mention to the document.